src/rules/noInternalModuleRule.ts

```diff
@@ -24,12 +24,32 @@ export class Rule extends Lint.Rules.AbstractRule {
 
 class NoInternalModuleWalker extends Lint.RuleWalker {
     public visitModuleDeclaration(node: ts.ModuleDeclaration) {
-        if (Lint.isNodeFlagSet(node, ts.NodeFlags.Namespace)) {
-            // ok namespace
-        } else if (node.name.kind === ts.SyntaxKind.Identifier) {
-            // for external modules, node.name will be a LiteralExpression instead of Identifier
+        if (this.isInternalModuleDeclaration(node)) {
             this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
         }
         super.visitModuleDeclaration(node);
     }
+
+    private isInternalModuleDeclaration(node: ts.ModuleDeclaration) {
+        // an internal module declaration is not a namespace or a nested declaration
+        // for external modules, node.name.kind will be a LiteralExpression instead of Identifier
+        return node.name.kind === ts.SyntaxKind.Identifier
+            && !(this.isNamespaceDeclaration(node) || this.isNestedDeclaration(node));
+    }
+
+    // todo: remove this method and replace it with only the flag check when https://github.com/Microsoft/TypeScript/issues/4436 is fixed
+    private isNamespaceDeclaration(node: ts.ModuleDeclaration): boolean {
```

Comment:
Sadly, TypeScript cannot infer the return type because of the underlying recursion. Since this method is flagged for removal I hope it's not a problem if it stays like this for now.